### PR TITLE
[Heading block] Add a wp-block-heading CSS class

### PR DESCRIPTION
**Pull request by adamziel**
## What?
This PR adds the `wp-block-heading` CSS class to the heading block.

It happens server-side so:

* The class couldn't be accidentally removed from the serialized markup
* There's no need to migrate the existing posts

## Why?
Without the classname, global styles can't distinguish between the `h1-h6` elements that belong to the heading block vs ones that don't. As a result, [the styles set on the `core/heading` block of theme.json gets applied to all h1, h2, h3, etc](https://github.com/WordPress/gutenberg/issues/42082)

## How?

This PR transforms the stored `heading` block markup into one with an additional class name using  [WP_HTML_Tag_Processor](https://make.wordpress.org/core/2022/08/19/a-new-system-for-simply-and-reliably-updating-html-attributes/). 

## Alternatives considered

I explored leaning on `get_block_wrapper_attributes`, but it falls short:

* It infers attributes from globals.
* It only processes the `style` and `class` properties, meaning a heading block with an anchor set would lose its `id`.
* It ignores any additional attributes inside of the block content.
* We'd still need to some parsing to remove the wrapping block from `$content`.

## Testing Instructions

1. Create a new page
2. Add a few heading blocks
3. Align text to right in one of them
4. Add a custom CSS classname in another
5. Save
6. Visit the page, confirm the rendered headings have the `wp-block-heading` css class

cc @draganescu @getdave @scruffian

### CODE REVIEW

**Comment by spacedmonkey** [comment]
@adamziel Regex is express and error prone. Can we add a couple of unit tests please?

**Comment by scruffian** [comment]
This looks good and works well. The only thing I might change is to remove this leading space, but it's not a big deal:
Screenshot 2022-07-04 at 14 11 02

**Comment by adamziel** [comment]
@spacedmonkey I've just added a bunch of tests!

@scruffian I got rid of that space. Unfortunately, it required switching from `preg_replace` to `preg_match` and manual processing. Trade-offs, trade-offs 🤷

File: phpunit/blocks/render-block-heading-test.php
Comment (spacedmonkey) [inline phpunit/blocks/render-block-heading-test.php]:
```suggestion
			array(
				'<span>Hello World</span>',
				'<span>Hello World</span>'
			),
			array(
				'<h7>Hello World</h7>',
				'<h7>Hello World</h7>'
			),
```

Let's add a test for h7, as an invalid header tag.

File: phpunit/blocks/render-block-heading-test.php
Comment (spacedmonkey) [inline phpunit/blocks/render-block-heading-test.php]: I ❤️ the data provider. Can we add array keys, so it easier to debug on test failure. See this example.

https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/tests/phpunit/tests/rest-api/application-passwords.php#L114-L124

File: packages/block-library/src/heading/index.php
Comment (spacedmonkey) [inline packages/block-library/src/heading/index.php]: Is it worth adding this here. 

```
if ( empty( $content ) ) {
		return $content;
	}
```

Comment (spacedmonkey) [reply packages/block-library/src/heading/index.php]: or 
```
if( str_contains( 'wp-block-heading', $content ){
   return $content;
}
```

File: phpunit/blocks/render-block-heading-test.php
Comment (spacedmonkey) [inline phpunit/blocks/render-block-heading-test.php]: Let's ensure that class is not double added. 

```
			array(
				'<h2 class="is-align-right wp-block-heading">Hello World</h2>',
				'<h2 class="is-align-right wp-block-heading">Hello World</h2>'
			),
```

**Comment by spacedmonkey** [comment]
Unit tests are very useful. Added some comments. 

I really hope there unit tests are ported to core on core merged.

File: packages/block-library/src/heading/index.php
Comment (draganescu) [inline packages/block-library/src/heading/index.php]: Possible dumb question: should we not add this class name if it already exists?

File: packages/block-library/src/heading/index.php
Comment (draganescu) [inline packages/block-library/src/heading/index.php]: I believe we should do this for all blocks and automagically add the `wp-block-{blockName}` class when rendering the pages on the server.

**Comment by markhowellsmead** [comment]
<s>Using PHP `DOMDocument` to parse the block's attribute may be more readable and reliable. I haven't noticed any performance issues when I've used it.</s>
Never mind, I see that's been discussed in https://github.com/WordPress/gutenberg/pull/40859 and that it might introduce other problems.

**Comment by Mamaduka** [comment]
>It does so server-side so that the class is not stored in the serialized markup where it could be accidentally removed.

If you try removing the block class name via Code Editor, it will be added back, saving the content.

@adamziel, any reason for not using block deprecations for this? I know deprecation won't update content without resaving it, but this is how we usually update the block markup.

P.S. The Heading block has `className` support explicitly disabled.

https://github.com/WordPress/gutenberg/blob/59fb859b08f5f6dfd40a829818682f7e2691d45e/packages/block-library/src/heading/block.json#L29-L32

File: packages/block-library/src/heading/index.php
Comment (spacedmonkey) [reply packages/block-library/src/heading/index.php]: See https://github.com/WordPress/gutenberg/pull/42122#discussion_r913081338

**Comment by adamziel** [comment]
The class name needs to be applied both in the editor and on the backend, or else the following mismatch will happen:

style.scss:
```scss
.wp-block-heading {
	background-color: #0c88b4;
}
```

theme.json:
```js
{
	"styles": {
		"blocks": {
			"core/heading": {
				"color": {
					"background": "pink"
				}
			}
		}
	}
}
```

Editor view, where the CSS `wp-block-heading` class is not applied:
CleanShot 2022-07-05 at 16 48 56@2x

Website view, where the CSS `wp-block-heading` class is applied:
CleanShot 2022-07-05 at 16 49 07@2x

> If you try removing the block class name via Code Editor, it will be added back, saving the content.

@Mamaduka This PR would only add it when rendering, not when saving the content. Exploring an on save backend hook would be an interesting option, though.

> any reason for not using block deprecations for this?

It wouldn't add the class to the blocks that already saved, meaning the styles wouldn't work correctly across your site until you manually migrate all your existing blocks :(

> P.S. The Heading block has className support explicitly disabled.

This is great for the editor!

File: packages/block-library/src/heading/index.php
Comment (adamziel) [reply packages/block-library/src/heading/index.php]: Good call!

File: packages/block-library/src/heading/index.php
Comment (adamziel) [reply packages/block-library/src/heading/index.php]: Agreed! Let's ship this PR and explore it in a follow-up

File: phpunit/blocks/render-block-heading-test.php
Comment (ZebulanStanphill) [inline phpunit/blocks/render-block-heading-test.php]: These comments need to be updated.

File: packages/block-library/src/heading/block.json
Comment (adamziel) [inline packages/block-library/src/heading/block.json:32]: Without this change, the class name won't be there in the editor.

**Comment by adamziel** [comment]
It looks like I'll need to revisit a few tests before this one can be merged.

**Comment by dmsnell** [comment]
Though it seems like a more difficult challenge in the short term working through deprecations and updating the block like @Mamaduka mentioned should have a good long-term impact.

> The class couldn't be accidentally removed from the serialized markup

This isn't a very helpful argument with blocks since _any_ markup changes can accidentally occur. If we accidentally change the block from `core/heading` to `kore/heading` then it's also destroyed. In other words, we can't really say we're guarded the block against accidental manual changes by spreading the logic for the block between the front and back ends. It's a good goal, but also one which can distract us on a long journey of futility.

Something that we might consider in a PR like this one is thinking more about updating the block going forward and providing a transition fix on the backend. Ultimately we shouldn't need that fix because the block markup is fixed where it's created.

But also, blocks that need updating are a big problem. Maybe we can also scope-limit changes like these, which add code we may never get to remove even when we want to, and use the need to prioritize updating/migrating blocks outside of edit sessions.

Correct me if I'm wrong, but deprecations occur for a full post at a time. We don't have to edit a specific block in order to migrate it; we only need to make _any_ edit on _any_ block within the same post that it's in… So I wonder if the case in your screenshot with _some_ headings properly colored and other not is a likely scenario or an exception to the rule.

**Comment by adamziel** [comment]
> Though it seems like a more difficult challenge in the short term working through deprecations and updating the block like @Mamaduka mentioned should have a good long-term impact.

We can never assume that all the headers in all WordPresses have already been migrated, therefore the fallback will always be needed.

That being said, I explored the migration more.

The bad news: The deprecations API won't help here as the `isEligible` function decides based on block attributes and innerBlocks, none of which are getting changed.

The good news: The editor is doing the migration automatically behind the scenes and even handles the "dirty" state correctly – which is what I was worried about the most. As soon as I switch `supports.className` to true and open a page with a pre-existing class-less heading block, the editor updates the serialized markup.

The original block markup:

```html
<!-- wp:heading -->
<h2>heading</h2>
<!-- /wp:heading -->
```

The block markup after refreshing the editor with `supports.className == true`:

```html
<!-- wp:heading -->
<h2 class="wp-block-heading">heading</h2>
<!-- /wp:heading -->
```

Which is to say, the migration is baked into this PR 👍 @Mamaduka @dmsnell and also cc @mtias

**Comment by dmsnell** [comment]
> We can never assume that all the headers in all WordPresses have already been migrated, therefore the fallback will always be needed.

there's always [a reasonable cutoff](https://core.trac.wordpress.org/ticket/33848).

**Comment by adamziel** [comment]
> Netscape 4

That's a browser I haven't seen in a long time indeed! Still, Netscape was a matter of measurable market share while the header tags are a matter of unmeasurable amount of still unmigrated tags out in the wild. Right now I don't see a way out of having that fallback code baked-in even ten years from now, but I'm happy to revisit in a few years 🤷

File: packages/block-library/src/heading/index.php
Comment (adamziel) [inline packages/block-library/src/heading/index.php]: I assumed double quotes here since that's how Gutenberg serializes block markup. Happy to iterate on that if needed.

**Comment by adamziel** [comment]
This one is ready for a re-review! @draganescu @spacedmonkey @dmsnell @ZebulanStanphill @getdave @scruffian @Mamaduka

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [reply packages/block-library/src/heading/index.php]: if we're going to stop and acknowledge we're making this assumption why not avoid the need to make it?

```
class=(?P<quote>['"]).*?(\k{quote})
             |         |   \- until we find that quote character again
             |         \- un-greedy match of any character
             \- capture the quote character
```

Comment (adamziel) [reply packages/block-library/src/heading/index.php]: Good idea! I just updated this PR

File: packages/block-library/src/heading/index.php
Comment (adamziel) [reply packages/block-library/src/heading/index.php]: The follow-up: https://github.com/WordPress/gutenberg/pull/42269

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: this isn't particularly doing server-side rendering so much as it's modifying the existing markup.

of note, we don't have anywhere in this file that paints a picture of what we're trying to do. the closest thing to it is inside the PCRE pattern and that's not too clear. an example showing the transformation would be a major aid to whoever has to come back here in the future, figure out what this does, figure out what it was supposed to do instead, and figure out how to try and fix it.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: these negative lookahead can behave very slowly in practice, particularly if we end up in backtracking situations. ungreedy matches can help us because they let us express everything _until_ this part that we're looking for but don't incur the same performance issues that negative lookahead do

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: All of these assignments by parts is a bit hard to follow, plus it feels like we're running around an awkward way to do string concatenation.

```php
$new_tag = "<{$matches['tag_name']}>{$matches['before_class']} class={$quote}{$new_class_names}{$quote}{$matches['after_class']}>;
```

Note we shouldn't need to worry about trimming the attributes either because extra space between attributes and class names is irrelevant. The same is true when adding the new class name to the list; we also don't have to worry about adding a class twice since that too should be irrelevant.

```php
$new_class_names = "{$matches['class_name']} wp-block-heading";
```

It should work fine without even checking if `$matches['class_name']` is empty or not because if it's not there the matched value should be the empty string.

a few of these changes should reveal a simpler function here.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: it looks like this assumes that nothing else can precede our match other than whitespace, also that we're intentionally applying this to the first HTML tag whatever it is. that's fine, if that's the case. just pointing out the assumption implicit in this substitution.

remember we also have `preg_replace_callback()` which can handle the replacement for us

File: packages/block-library/src/heading/index.php
Comment (adamziel) [reply packages/block-library/src/heading/index.php]: I know how to do negative non-greedy matching with character classes, but how would you do it with a word?

Comment (adamziel) [reply packages/block-library/src/heading/index.php]: Nevermind, it clicked the moment I pressed "submit".

**Review by draganescu — APPROVED** [review]
I think @dmsnell 's [comment](https://github.com/WordPress/gutenberg/pull/42122/files#r916821920) on how we compose the new tag is on point but the PR is good for approval IMO

**Comment by adamziel** [comment]
@dmsnell thank you for great notes, the function just got much simpler!

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php:15]: worth gold ✨

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: I think the `\b` is not a blocker here, but just so we're aware, it captures probably more than we want, including things like punctuation.

also fun fact! PHP has supported list restructuring for a loooong time

```php
list( '0' => $fullmatch, 'quote' => $quote, 'class_name' => $class_name ) = $matches;
```

I like using this as an aid to comprehension though if a capture group is optional you have to either tell the PCRE engine to include it anyway _or_ grab that capture group like normal outside of `list()` where we can `if ( isset( ) )` it

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: I believe that this line is wholly redundant, because if there's nothing inside the `class_name` match the value of `$matches['class_name']` will be `''`, so we can just use it as-is.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: I suggested `(?P<quote>…)` but you already used named capture group syntax without the `P`. Probably worth only using one flavor of the syntax so that it doesn't appear like we intentionally chose one or the other for each case.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: like my comment above, I believe we've constructed the pattern in such a way that if the quote isn't there the pattern fails, so `$matches['quote']` will always be non-empty. we can assume it's there because it will be by construction

File: phpunit/blocks/render-block-heading-test.php
Comment (draganescu) [inline phpunit/blocks/render-block-heading-test.php:66]: What is the double space here?

Comment (dmsnell) [reply phpunit/blocks/render-block-heading-test.php:66]: looks like an over-specified test, making assertions about inconsequential elements of the unit under test.

**Comment by drdogbot7** [comment]
I love this, but can we and add classes for specific levels of heading too?

```html
<h1 class="wp-block-heading wp-block-heading--h1">Level One Heading</h1>
<h2 class="wp-block-heading wp-block-heading--h2">Level One Heading</h2>
<h3 class="wp-block-heading wp-block-heading--h3">Level One Heading</h3>
<h4 class="wp-block-heading wp-block-heading--h4">Level One Heading</h4>
<h5 class="wp-block-heading wp-block-heading--h5">Level One Heading</h5>
<h6 class="wp-block-heading wp-block-heading--h6">Level One Heading</h6>
```

The more we can get away from using classes instead of global tags in our CSS the easier it will be to work with the block editor.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php]: this is an example of where we don't necessarily want to require a fixed tag name. it may be the _only_ example because h1-h6 are special, whereas there tags are governed by attributes and not by their tag name (to determine the header level I mean).

we _could_ find the first tag, or find _any_ tag if it has an appropriate class name on it already. or we could expand the matcher modifier to allow _some_ form of matching multiples, or add a special-casing in the matcher to identify headings…

**Comment by drdogbot7** [comment]
> > any reason for not using block deprecations for this?
> 
> It wouldn't add the class to the blocks that already saved, meaning the styles wouldn't work correctly across your site until you manually migrate all your existing blocks :(

How big an issue is that _really_ though? These classes are largely additive and I don't see how adding them to new headings would **break** existing sites.

Yes, you would have problems if (1) you have a site with legacy content and (2) you switch to a theme that relies on these heading classes being there.

Even with this PR though, without a fallback, you'll still have issues with any headings inside a classic block or created with the classic editor:

image

I've always assumed supporting the classic editor was the reason we didn't get classes on common elements in the first place.

We absolutely should have classes on common block elements, for simplicity's sake, I think it makes sense to just use block deprecation. Public/premium themes are going to need to provide fallbacks for the foreseeable future anyway.  New sites with custom themes can start using these right away.

**Comment by gaambo** [comment]
I think this PR is a great example of how we have to use complex regex to do something that should be simple: Adding/modifying a HTML attribut to block markup. Those regex have to be documented and tested very well. 
Is that the best way to do it? Would DOMDocument be better - or can we think of other ways to handle similar things in the future?

**Comment by dmsnell** [comment]
@gaambo: @adamziel and I have been working on some ideas for a new API to modify HTML to make these kinds of additions trivial. we discussed a lot in person and so we still have a lot of basic documentation of the idea to do and we're both prototyping a solution to make sure it's viable, but as block or plugin authors we could write stuff like this and leave it to Core to do the changes.

```php
function block_core_heading_render( $attributes, $content ) {
 	if ( ! $content ) {
 		return $content;
 	}

	return ( new WP_HTML_Processor( $content ) )
		->find( 'h' ) // bad example maybe because h1 - h6 are special
		->add_class( 'wp-block-heading' );
}
```

The secret sauce is that we're avoiding a full HTML parse of the block HTML and instead are mostly just parsing the HTML opening tags, which we can do more reliably.

Unfortunately, despite being a personal advocate of `DOMDocument` it seems to be insufficient for Core work because it's not always available and because it has some strange quirks, especially when provided invalid HTML. It would probably be good enough if it weren't for the the fact that it's hard to ensure all or even most installs have that extension available, and if we want to provide a fallback for unsupported servers, we might as well focus on that fallback and not require the extension.

See #42485 and #42507

**Comment by adamziel** [comment]
With https://github.com/WordPress/gutenberg/pull/42485#issuecomment-1255937246 merged, this is now unblocked 🎉

**Comment by carolinan** [comment]
What is the next step for this CSS class change? Is it blocked or not?

**Comment by getdave** [comment]
> What is the next step for this CSS class change? Is it blocked or not?

We need to change the implementation in this PR to not use regex for replacement. Rather it can now use the HTML Tag Parser that @adamziel and @dmsnell have been working on.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [inline packages/block-library/src/heading/index.php:38]: this function is about to get rather boring. I think this is right, but please check me on it

```php
$tags = new WP_HTML_Tag_Processor( $content );

$header_tags = array( 'H1', 'H2', 'H3', 'H4', 'H5', 'H6' );
while ( $tags->next_tag() ) {
	if ( in_array( $tags->get_tag(), $header_tags, true ) ) {
		$tags->add_class( 'wp-block-heading' );
	}
}

return $tags->get_updated_html();
```

File: packages/block-library/src/heading/index.php
Comment (adamziel) [inline packages/block-library/src/heading/index.php:32]:
```suggestion
			$p->add_class( 'wp-block-heading' );
			break;
```

**Comment by adamziel** [comment]
I'd like to ship this PR soon. I rebased, adjusted most test snapshots, and it's down to adjusting the Playwright E2E tests which is next on my list.

File: packages/block-library/src/heading/index.php
Comment (dmsnell) [reply packages/block-library/src/heading/index.php:32]: so I believe that originally, as written, the function modified all heading tags, something which I tried to reproduce in my suggested code. `preg_replace_callback` will continue to scan unless given a `$limit` parameter.

glad this is clearer now what the behavior should be.

**Comment by adamziel** [comment]
I'll go ahead and merge – rebasing is painful and the PR seems ready.

The tests are passing, the approval is there from back when it was in worse shape due to regular expressions, and the only piece of feedback has been addressed by replacing the custom regexps with  `WP_HTML_Tag_Processor` code. 

If anything is badly off, I'm happy to promptly follow-up or even revert if needed.

**Comment by aaronrobertshaw** [comment]
>  any reason for not using block deprecations for this?
>
> It wouldn't add the class to the blocks that already saved, meaning the styles wouldn't work correctly across your site until you manually migrate all your existing blocks :(

@adamziel without adding a deprecation, preexisting heading blocks will throw a block validation error. This is now what I'm seeing on trunk due to heading blocks both in patterns and existing content.

Screenshot 2022-11-29 at 10 16 34 am

Checking out the commit prior to this one shows those errors disappear.

> If anything is badly off, I'm happy to promptly follow-up or even revert if needed.

Given those errors, I'd say we need a deprecation or temporary revert.

File: packages/block-library/src/heading/block.json
Comment (andrewserong) [inline packages/block-library/src/heading/block.json:32]: It looks like this change might also mean that the className is being serialized to post content. Was that intentional?

**Comment by aaronrobertshaw** [comment]
I've created a PR adding the missing deprecation and fixtures: https://github.com/WordPress/gutenberg/pull/46138.

If the serialization of the classname was deliberate, as the updates to the existing fixtures suggest, that PR should help sort things out.

**Comment by adamziel** [comment]
@andrewserong @aaronrobertshaw Oh that’s a good point! I know @mtias wanted to avoid having the class name in the serialized markup, and at the same time I remember there were some challenges related to that. I’ll revisit this choice when I’m at the computer today. If this one derails dev envs, feel free to revert.

**Comment by adamziel** [comment]
> > It looks like this change might also mean that the className is being serialized to post content. Was that intentional?
> I know @mtias wanted to avoid having the class name in the serialized markup, and at the same time I remember there were some challenges related to that. I’ll revisit this choice when I’m at the computer today.

The complication is [how setting `supports.className = true` in block.json works](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/):

> By default, the class .wp-block-your-block-name is added to the root element of your saved markup. This helps having a consistent mechanism for styling blocks that themes and plugins can rely on. If, for whatever reason, a class is not desired on the markup, this functionality can be disabled.

The class name could be accidentally removed, but that shouldn't be a problem as @Mamaduka explained earlier in this thread:

> If you try removing the block class name via Code Editor, it will be added back, saving the content.

To have the className without storing it in the markup, we'd have to invent a new supports API that would:

* Add the class name in the editor view, but not in the code view
* Add the class name server-side when rendering the markup
* Make double sure the class name is not being added to the markup

I'd rather not do that just for the header block – it adds complexity to the code and makes the supports harder to understand. If it deems indispensable for any reason, we could consider adding it later on alongside another deprecation. Let's stick to storing the `wp-block-heading` class in the markup for now. CC @mtias 

As for adding a `wp-block-*` class to every block, let's discuss that separately.

File: packages/block-library/src/heading/block.json
Comment (adamziel) [reply packages/block-library/src/heading/block.json:32]: See https://github.com/WordPress/gutenberg/pull/42122#issuecomment-1331479800

**Comment by andrewserong** [comment]
> If it deems indispensable for any reason, we could consider adding it later on alongside another deprecation. Let's stick to storing the wp-block-heading class in the markup for now.

Sounds good to me, thanks for following up! Looks like it'll be good to land https://github.com/WordPress/gutenberg/pull/46138 then 👍

**Comment by adamziel** [comment]
Dev note for 6.2  cc @bph @MaggieCabrera  @andrewserong :

## WordPress 6.2 adds a wp-block-heading CSS class to every Heading Block

In WordPress 6.2, the h1-h6 elements added via the heading block have a brand new wp-block-heading CSS class.

This change enables styling the heading block differently from the regular h1-h6 elements. For example, the following theme.json would add a blue background to all h1 elements and a pink background to only the h1 elements added via the heading block:

```json
{
	"styles": {
		"elements": {
			"h1": {
				"color": {
					"background": "blue"
				}
			}
	        },
		"blocks": {
			"core/heading": {
				"elements": {
					"h1": {
						"color": {
							"background": "pink"
						}
					}
				}
			}
		}
	}
}
```

Work is underway to add a CSS class to every Gutenberg block in future WordPress releases.

Props to @ajlende for reviewing and editing.

**Comment by bph** [comment]
Thank you, @adamziel. [Added Dev Note to Misc Editor post. ](https://make.wordpress.org/core/?p=102924&preview=1&_ppp=1f21a08839)

**Comment by sabernhardt** [comment]
The dev note example is not working for individual heading elements with the block's class. [See Trac 57904](https://core.trac.wordpress.org/ticket/57904). Is it just a mistake in the example?

**Comment by scruffian** [comment]
Since this will work for any block except headings, I suggest we change the dev note to use a different block.

**Comment by scruffian** [comment]
Let's use this example for the dev note:
```
{
	"styles": {
		"elements": {
			"h1": {
				"color": {
					"background": "blue"
				}
			}
	        },
		"blocks": {
			"core/cover": {
				"elements": {
					"h1": {
						"color": {
							"background": "pink"
						}
					}
				}
			}
		}
	}
}
```

This will mean that H1 elements inside a cover block will have a pink background, but all other H1s will have a blue background.

**Comment by bph** [comment]
[Just updated the dev note. ](https://make.wordpress.org/core/2023/03/07/miscellaneous-editor-changes-in-wordpress-6-2/#wp-block-heading-css-class-added-to-every-heading-block)

[referenced in trac, too](https://core.trac.wordpress.org/ticket/57904#comment:14)

**Comment by adamziel** [comment]
I've dropped the ball here, thank you so much @scruffian!

**Pull request merged**
